ide/coqide/coq.ml

```diff
@@ -672,7 +672,7 @@ struct
     { opts = [universes]; init = false; label = "Display _universe levels" };
     { opts = [all_basic;existential;universes]; init = false;
       label = "Display all _low-level contents" };
-    { opts = [unfocused]; init = false; label = "Display _unfocused goals" };
+    { opts = [unfocused]; init = false; label = "Display un_focused goals" };
```

Comment:
Tangent: the underline navigation mechanism is really weird.

![image](https://user-images.githubusercontent.com/1253341/230509046-b2c4f143-98d2-455b-9a6b-c20055e366e5.png)

The letter doesn't have to be unique.  Without the change, you'd just press `u` twice to get to the entry.  OTOH, I don't see how this form of menu navigation is all that useful:
- Getting the underlining to show in the menu is quirky.  AFAICT, you have to click on a top menu, then type a letter.  If the letter isn't assigned (e.g. `a`), they won't appear unless you navigate with a right/left arrow (among other quirks).
- You can't activate these directly from the editor panel or, e.g., the Messages panel (read-only).
- The functionality mostly duplicates hot key sequences (e.g. Shift+Alt+F)
- It's hard to pick out the underlined characters.

I wonder if there's a way to disable it--though I'd ask others' opinions first.

Reply:
> The letter doesn't have to be unique.

You should open the current CoqIDE version and see what it does: this mnemonic letter is also used to determine the accelerator (default is: Shift-Alt-U, if U is chosen as a mnemonic) actually. See:
https://github.com/coq/coq/blob/584a762bd4c2a6a5d04c0c71246b7130b80d618e/ide/coqide/coqide.ml#L1271
Furthermore, for efficiency, it’s still better to choose distinct mnemonics in a menu item. I think it has importance as most GUI pieces of software still provide mnemonics. For instance, in Ms Office with the ribbon, you can still access the ribbon’s elements using Alt-letter, letter, letter, and they sometimes mention it in help contents if I remember well. Mnemonics also have historical importance: when GUIs appeared, while mouse devices were quite rare, mnemonics were the common way to quickly access menu items. Personally, I do use mnemonics for some items for which there isn’t a shortcut, notably in context menus: right-click, letter; or on buttons: Alt-letter; and my action is done in a fraction of a second without any effort. It is also helpful if the existing accelerator is borrowed by the desktop environment, for example.

To summarize: mnemonic letters are a very standard, easy-to-deal-with, non-bothering feature of GUIs. I hope this is enough to convince you :slightly_smiling_face:.

> It's hard to pick out the underlined characters.

At least, it has no cost to provide such mnemonics, no matter whether the choice is definitely the best or not, so let’s do it.

> Getting the underlining to show in the menu is quirky. / You can't activate these directly from the editor panel or, e.g., the Messages panel (read-only).

I don’t see what you mean; you may have behaviors which are specific to your desktop environment. Formerly (i.e. before Windows 95), mnemonics were shown all the time. Strangely, it seems to be the case in my Firefox and Thunderbird too. The standard way to use mnemonics is to press the underlined letter while holding Alt. Therefore, the underlining only appears when you press Alt. Historically, you could also press F10 to activate the menu bar, and Shift-F10 to show the context menu (this does not work in CoqIDE because these key shortcuts are borrowed by Debug menu items); now, there is the special Menu key for the context menu. Under Windows, you may also press Alt, letter instead of Alt-letter (meaning you can press Alt then release it, then press the letter, instead of pressing both at the same time), and F10 does not open the first menu (usually “File”) but only selects it.

When I was a child, my first feeling about these mnemonics was that they were not so useful, too. But then I discovered the option in the so-called “Folder options” on the Windows explorer: “Do not show rapid access keys before I press Alt” (I think it was disabled by default before Windows XP). Indeed, mnemonics are much more useful when you know they work with Alt :slightly_smiling_face:.

In the “Find in page” bar of Firefox, some time ago you had “_Previous” and “_Next” buttons, so you could press Alt-P to “Find previous.” Now that they’ve been replaced by buttons with icons, I don’t know if there is a shortcut for this and it’s annoying.

The common character to put before mnemonics (in other GUI libraries) is the ampersand `&`, but GTK requires an underscore instead, I don’t know exactly why.

Reply:
I've never used or wanted such mnemonics.  Only accelerators, which I use sparingly.  Too easy to pick the wrong operation.

> I don’t see what you mean; you may have behaviors which are specific to your desktop environment.

The documentation should say something about how to access mnemonics (e.g. to press Alt; I was opening a menu and then pressing left/right to see the mnemonics).  Also the convention for how CoqIDE assigns mnemonics and accelerators. (I hope there is an underlying consistency.  What we have seems haphazard.)  Contrary to what many believe, GUIs are not self-documenting.  The behavior of mnemonics is not simple to discover.

I think linking the accelerator to the mnemonic is problematic.  And it's too easy to bind an accelerator to multiple items (e.g. for CTRL); there should be a runtime error for that.  Nor do I see underscore for the File menu entries.

```
  menu file_menu [
    item "File" ~label:"_File";
    item "New" ~callback:File.newfile ~stock:`NEW;
    item "Open" ~callback:(File.load ~parent:w) ~stock:`OPEN;
```

But let's get back to the task at hand.

Reply:
> I was opening a menu and then pressing left/right to see the mnemonics)

Okay. I thought I tried this and it didn’t work… But indeed it works. But it is not very useful if you have to use your mouse first.

> The documentation should say something about how to access mnemonics (e.g. to press Alt

We can mention it somewhere if you want. But please keep in mind that this is really a standard. By this, I mean: just as when you learn that Shift is to write capital letters and characters at the top of keys, that AltGr is for  or that Ctrl is for main hotkeys, or that there are navigation keys (too few people know about Home, End, PgUp, PgDn), you should learn that Alt is primarily for mnemonics. CoqIDE is not responsible here; the problem is that information on these standards hasn’t spread yet. People have to be more serious when teaching how to use a computer. When you have these keys so many hours in a week, it is important to know what they are useful for!